### PR TITLE
Adding a way to ignore specific LogGroups in check client compatibility

### PR DESCRIPTION
**Pull request by 09harsh**
# Description

*Please include a summary of the change and which issue is addressed. If this change resolves an issue, please include the issue number in the description.*

Fixes: #1512 

## Type of change

*Please select one of the options below.*

- New feature (non-breaking change which adds functionality)

## Discussion

*Are there any design details that you would like to discuss further?*
No

## Testing

*Please describe the tests that you ran to verify your changes. Unit tests?
Manual testing?*

*Do we need to perform additional testing once this is merged, or perform in a larger testing environment?*

Added unit tests.

## Documentation

*Did you update relevant documentation within this repository?*

*If this change is adding new functionality, do we need to describe it in our user manual?*

*If this change is adding or removing subreconcilers, have we updated the core technical design doc to reflect that?*

*If this change is adding new safety checks or new potential failure modes, have we documented and how to debug potential issues?*

## Follow-up

*Are there any follow-up issues that we should pursue in the future?*

*Does this introduce new defaults that we should re-evaluate in the future?*

### CODE REVIEW

**Review by johscheuer — CHANGES_REQUESTED** [review]
Please run `make clean all`

File: api/v1beta2/foundationdbcluster_types.go [hunk 1]
```diff
@@ -918,6 +918,9 @@ type FoundationDBClusterAutomationOptions struct {
 
 	// MaintenanceModeOptions contains options for maintenance mode related settings.
 	MaintenanceModeOptions MaintenanceModeOptions `json:"maintenanceModeOptions,omitempty"`
+
+	//IgnoreLogGroupsForUpgrade defines the list of LogGroups that should be ignored during fdb version upgrade.
```
Comment (johscheuer) [inline api/v1beta2/foundationdbcluster_types.go:922]:
```suggestion
	// IgnoreLogGroupsForUpgrade defines the list of LogGroups that should be ignored during fdb version upgrade.
```

Can we add a limitation to the max items of this slice?

**Review by johscheuer — CHANGES_REQUESTED** [review]
Could you add please test cases? And extend what kind of testing you did?

File: config/crd/bases/apps.foundationdb.org_foundationdbclusters.yaml [hunk 2]
```diff
@@ -1,9 +1,10 @@
+
 ---
 apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.9.2
+    controller-gen.kubebuilder.io/version: v0.6.1
```
Comment (09harsh) [inline config/crd/bases/apps.foundationdb.org_foundationdbclusters.yaml:7]: this seems wrong 🤔 
I am not sure why kubebuilder version was downgraded after make clean all.

File: controllers/check_client_compatibility_test.go [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+			func(tc testCase) {
+				unsupportedClients := getUnsupportedClients(supportedVersion, "fdb00b061060001", tc.ignoredLogGroups)
+				Expect(unsupportedClients).To(ContainElements(tc.expectedUnsupportedClients))
+				Expect(unsupportedClients).To(HaveLen(len(tc.expectedUnsupportedClients)))
+			},
+			Entry("With empty IgnoreProcessGroups list.",
+				testCase{
+					ignoredLogGroups: nil,
+					expectedUnsupportedClients: []string{
+						"10.1.38.106:35640 (sample-cluster-client)",
```
Comment (johscheuer) [inline controllers/check_client_compatibility_test.go:210]: Can we change the input data to not have the same log group for all reported clients?

File: controllers/check_client_compatibility_test.go [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+				LogGroup: "sample-cluster-client",
+			},
+		},
+		ProtocolVersion: "fdb00b062010002",
+		SourceVersion:   "20566f2ff06a7e822b30e8cfd91090fbd863a393",
+	},
+}
+
+var _ = Describe("check client compatibility", func() {
+	When("validating getUnsupportedClient()", func() {
```
Comment (johscheuer) [inline controllers/check_client_compatibility_test.go:194]:
```suggestion
	When("getting the list of unsupported clients from the cluster status json", func() {
```

File: controllers/check_client_compatibility_test.go [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+			expectedUnsupportedClients []string
+		}
+
+		DescribeTable("should return all the unsupported clients",
+			func(tc testCase) {
+				unsupportedClients := getUnsupportedClients(supportedVersion, "fdb00b061060001", tc.ignoredLogGroups)
+				Expect(unsupportedClients).To(ContainElements(tc.expectedUnsupportedClients))
+				Expect(unsupportedClients).To(HaveLen(len(tc.expectedUnsupportedClients)))
+			},
+			Entry("With empty IgnoreProcessGroups list.",
```
Comment (johscheuer) [inline controllers/check_client_compatibility_test.go:206]:
```suggestion
			Entry("with an empty ignoreProcessGroups list.",
```

File: controllers/check_client_compatibility_test.go [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+						"10.1.38.106:36128 (sample-cluster-client)",
+						"10.1.38.106:36802 (sample-cluster-client)",
+						"10.1.38.107:42234 (sample-cluster-client)",
+						"10.1.38.107:49684 (sample-cluster-client)",
+						"10.1.38.108:47320 (sample-cluster-client)",
+						"10.1.38.108:47388 (sample-cluster-client)",
+						"10.1.38.108:58734 (sample-cluster-client)",
+					},
+				}),
+			Entry("With non empty IgnoreProcessGroups list.",
```
Comment (johscheuer) [inline controllers/check_client_compatibility_test.go:220]:
```suggestion
			Entry("with non empty IgnoreProcessGroups map.",
```

File: controllers/check_client_compatibility_test.go [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+var _ = Describe("check client compatibility", func() {
+	When("validating getUnsupportedClient()", func() {
+		type testCase struct {
+			ignoredLogGroups           map[string]fdbv1beta2.None
+			expectedUnsupportedClients []string
+		}
+
+		DescribeTable("should return all the unsupported clients",
+			func(tc testCase) {
+				unsupportedClients := getUnsupportedClients(supportedVersion, "fdb00b061060001", tc.ignoredLogGroups)
```
Comment (johscheuer) [inline controllers/check_client_compatibility_test.go:202]:
```suggestion
				unsupportedClients := getUnsupportedClients(supportedVersion, "fdb00b063010001", tc.ignoredLogGroups)
```

Doing a check for the 6.1 version seems wrong.

File: controllers/check_client_compatibility_test.go [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+				Expect(unsupportedClients).To(ContainElements(tc.expectedUnsupportedClients))
+				Expect(unsupportedClients).To(HaveLen(len(tc.expectedUnsupportedClients)))
```
Comment (johscheuer) [inline controllers/check_client_compatibility_test.go:204]:
```suggestion
				Expect(unsupportedClients).To(ConsistOf(tc.expectedUnsupportedClients))
```

File: api/v1beta2/foundationdbcluster_types.go [hunk 9]
```diff
@@ -918,6 +918,10 @@ type FoundationDBClusterAutomationOptions struct {
 
 	// MaintenanceModeOptions contains options for maintenance mode related settings.
 	MaintenanceModeOptions MaintenanceModeOptions `json:"maintenanceModeOptions,omitempty"`
+
+	// IgnoreLogGroupsForUpgrade defines the list of LogGroups that should be ignored during fdb version upgrade.
+	// +kubebuilder:validation:MaxItems=10
+	IgnoreLogGroupsForUpgrade []string `json:"ignoreLogGroupsForUpgrade,omitempty"`
```
Comment (johscheuer) [inline api/v1beta2/foundationdbcluster_types.go:924]: In a follow up we should define a typed string for the log group to provide some validation patterns for those strings.

**Review by johscheuer — CHANGES_REQUESTED** [review]
Changes overall look good, only a few recommendations. Have you done any testing besides the unit tests?

**Comment by 09harsh** [comment]
> Changes overall look good, only a few recommendations. Have you done any testing besides the unit tests?

No apart from unit tests I haven't done any other testing.

File: api/v1beta2/foundationdbcluster_types.go (hunk 9 above)
Comment (johscheuer) [reply api/v1beta2/foundationdbcluster_types.go:924]: Could you create a GitHub issue for this?

**Review by johscheuer — APPROVED** [review]
LGTM, could you rebase your changes and create an issue?

**Comment by 09harsh** [comment]
> LGTM, could you rebase your changes and create an issue?

Sure

**Comment by 09harsh** [comment]
Created issue for the same: https://github.com/FoundationDB/fdb-kubernetes-operator/issues/1542

File: api/v1beta2/foundationdbcluster_types.go (hunk 9 above)
Comment (09harsh) [reply api/v1beta2/foundationdbcluster_types.go:925]: done

**Comment by 09harsh** [comment]
```
[Container] 2023/03/17 14:36:18 Running command if $(grep -q -- "--- FAIL:" logs/*.log); then echo "TESTS FAILED SEE THESE LOGS:"; echo ; grep -l -- "--- FAIL:" logs/*.log; export fail_test=true; fi
TESTS FAILED SEE THESE LOGS:

logs/test_operator_ha_upgrades.log
```

**Review by johscheuer — APPROVED** [review]
Waiting for the CI result. Once we have the result and it looks good, we can go ahead an merge it. Thanks!

**Pull request merged**